v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/schema/SchemaDiscovery.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (C) 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.source.reader.io.schema;
+
+import com.google.cloud.teleport.v2.source.reader.io.exception.exception.SchemaDiscoveryException;
+import com.google.cloud.teleport.v2.spanner.migrations.schema.SourceColumnType;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import javax.sql.DataSource;
+
+/** Discover Schema of Source Tables. */
+public interface SchemaDiscovery {
```

Comment:
Is there a use case when someone would not want to do a retry ? Can we expose only the Retriable interface and the degenerate case be no-retries?

Reply:
Actually the Schema Discovery Implementation wraps the retriable interface and does the necessary retries.
The outside code will always use `SchemaDiscovery` which never throes a retriable exception.

This was done to make the retry handling logic common and avoiding the need to implement the same for say PG in future.